AXX-crl-enhancements.md

```diff
@@ -0,0 +1,220 @@
+AXX: Certificate Revocation List Enhancements
+----
+* Author(s): Gregory Cooke (@gtcooke94, @gregorycooke)
+* Approver: TODO 
+* Status: TODO
+* Implemented in: C, Go
+* Last updated: 2023-07-25
+* Discussion at: TODO
+
+## Abstract
+
+gRPC has an MVP implementation for CRLs (Certificate Revocation Lists). We wish to provide a more generic interface. In addition, we will provide two concrete implementations of this interface similar to Credential Reloading - a static provider and a directory reloader.
+
+
+## Background
+
+[RFC 5280](https://datatracker.ietf.org/doc/html/rfc5280) defines the desired behavior for Certificate Revocation Lists (CRLs). gRPC has an MVP implementation of this RFC for CRL handling in several languages, each with their own flavors - for example, C++ uses OpenSSL, whereas Go uses an implementation written in Go (and the stack in Java is significatly different). There are mismatches and deficiencies between the current Go and C++ implementations. Providing CRL reloading as a feature allows for more internal cohesion with the ideas behind credential reloading.
+
+### Related Proposals:
+
+* [L46: New TLS Credentials API](https://github.com/grpc/proposal/pull/205)
+
+## Proposal
+We seek to provide an alternate option for users to customize CRL handling. In a similar way that `CredentialReloading` is implemented, we will add a `CrlProvider`` interface through which gRPC fetches CRLs. In gRPC we will implement two providers (these match what we provide for certificates as well):
+
+1. A `StaticCrlProvider` that takes a list of strings that represent CRL file content
+1. A `CrlReloaderProvider` that periodically and asynchronously reloads a directory of CRLs.
+
+Users can also implement the interface with whatever behaviors they specifically desire.
+
+Per RFC5280, during revocation checking one of three status should be returned - `RevocationUnrevoked`, `RevocationRevoked`, and `RevocationUndetermined`. The user must specify whether `RevocationUndetermined` should be treated as `RevocationRevoked` or `RevocationUnrevoked` (failing open vs. failing closed).
+
+Assumptions:
+1. There should be no direct linkage between this interface and the creation/distribution of CRLs. gRPC is a user of CRLs and credentials, not a PKI itself.
```

Comment:
I'd add some 'high level overview' section (maybe as a comment to a diagram?) saying that existing impls use map as an in-memory storage and a provider to update it using some IOs. PKIs are completely out of scope of this gRFC

Reply:
Also let's add an assumption that we can't guess CRL provider by a name of a file / metadata of strings. However if someone wants to rely on it they can do a specific impl

Reply:
Added the second

For the first, this document has a structure defined by a template. I could have a High Level Overview sub-section in the Proposal section.

> existing impls use map as an in-memory storage and a provider to update it using some IOs

I'm a little confused at this comment - the existing CRL impls don't do this?